### ISSUE TITLE
JIT: Allow users to register external device code?

### ISSUE THREAD

**Issue by leofang** [cupy/cupy#6663]
### Description

During the past course of exposing CUDA C++ APIs to CuPy JIT I've gradually realized a big missing piece, which is to allow users to "bring in" their device kernels to CuPy JIT with ease. Currently it requires to go through the entire JIT stack to
- provide the calling signature in Python
  - the signature could deviate from C++'s, as it's hard to deal with raw pointers in Python
- make the types recognized by JIT
- special-case the treatment in the Python AST -> CUDA C++ translation

I feel the already-endless chase of CUDA built-in C++ APIs alone is tough enough, not to mention if users have device kernels that they'd like to interact with from Python. Such examples (that would interest multiple parties) include
- Exposing [NVSHMEM device-side communication APIs](https://docs.nvidia.com/hpc-sdk/nvshmem/api/docs/introduction.html#gpu-initiated-communication-and-strong-scaling) (cc: @dalcinl)
- Exposing CUDA math lib device APIs (ex: cuFFTDx) (cc: @mnicely)

It'd be nice -- in the future -- to add a registration system for users to register external device code and include the needed headers in CuPy JIT kernels, though I do not have good advices for implementation 😅 

A point of reference is Numba which will add [similar support](https://github.com/numba/numba/pull/7621), but there they already allow users to register foreign C/C++ types (through ctypes if I am not mistaken), so I am not sure it's something we could learn from...

**Comment by asi1024** [comment]
@leofang Would the following interface be enough to call external device code?

```py
code = '''
__device__ float device_func(float* x) { ... }
'''

@cupyx.jit.rawkernel(preamble=[code])
def kern(array):
    ...
    f = cupyx.jit.get_function_from_preamble('device_func', return_type=numpy.float32)
    y = f(array.data.ptr)  # No typecheck for arguments
    ...
```

**Comment by leofang** [comment]
Thanks, @asi1024. I will do a case survey and get back to you. My current concerns would be
- Don't we need to at least declare the number of inputs?
- Without type-checking input parameters, do we know how to generate the right code?

**Comment by asi1024** [comment]
I am considering the following implementation.
```py
class ExternalFunc(BuiltinFunc):
    def __init__(self, name, return_type='void'):
        self.name = name
        self.return_type = to_ctype(return_type)  # to_ctype is not implemented currently

    def call(self, env, *args):
        params = ', '.join([str(Data.init(x)) for x in args])
        return Data(f'{self.name}({params})', self.return_type)

def get_function_from_preamble(name, return_type):
    return ExternalFunc(name, return_type)
```

The types of input parameters are not checked unless explicitly checked in BuiltinFunc.call.

**Comment by kmaehashi** [comment]
Maybe we should also allow users to specify options to be passed to NVRTC (c.f. #6670)

**Comment by leofang** [comment]
> Maybe we should also allow users to specify options to be passed to NVRTC (c.f. #6670)

Yes, I think it was listed as a TODO in #5280 (CuPy JIT specific interface), also https://github.com/cupy/cupy/issues/4290#issuecomment-812317043 has similar asks. Probably a good time to design the interface for `jit.rawkernel()` 🙂

> The types of input parameters are not checked unless explicitly checked in BuiltinFunc.call.

I am still not sure I follow, sorry! For example, if I call an external kernel with `f(arr.data.ptr)`, how does JIT know if `arr.data.ptr` should be cast to `void*` or `void**` or any other pointer types?

> I will do a case survey and get back to you.

I did a quick survey. It should work for NVSHMEM because the device functions have C-like APIs. However, cuFFTDx is more challenging because it does not have clean separation of device/host APIs. I will submit feedbacks to the team but it should not be a blocker of this discussion, as in any case we need either spaghetti code or some new design to help JIT cuFFTDx code.

**Comment by leofang** [comment]
Adding a quick note here for future reference.

A recent offline discussion with the team suggests that [ctypes](https://docs.python.org/3/library/ctypes.html) based type declaration would be favored. Using the cuFFT new callback (#8242) as example:
```python
from ctypes import *

class cb_params(Structure):
    _fields_ = [("window_N", c_uint),
                ("signal_size", c_uint)]

class cufftComplex(Structure):
    _fields_ = [("x", c_float),
                ("y", c_float)]

@cupyx.jit.rawkernel(
    device=True,
    name='cufftJITCallbackLoadComplex',
    args=[POINTER(cufftComplex), c_size_t, POINTER(cb_params), POINTER(c_void_p)],
    returns=cufftComplex
)
def my_callback_func(input, index, info, sharedmem):
    if (index % info.signal_size) < info.window_N:
        return input[index]
    else:
        return cufftComplex(0, 0)
```
which can be converted to/from NumPy [(structured) dtypes](https://numpy.org/doc/stable/user/basics.rec.html), taking the `cb_params` type above as example:
```python
>>> import numpy as np
>>> cb_params_dtype = np.dtype([("window", np.uint), ("size", np.uint)], align=True)
>>> cb_params_dtype
dtype([('window', '<u8'), ('size', '<u8')], align=True)
>>> cb_params = np.ctypeslib.as_ctypes_type(cb_params_dtype)
```
then once CuPy supports NumPy structured dtypes (https://github.com/cupy/cupy/issues/2031) it’d work seamlessly on both host and device.

**Comment by dalcinl** [comment]
I would rather say that once #2031 gets implemented, you could just use NumPy dtypes for expressing kernel args and returns, and forget about ctypes. I'm biased here, I do not like `cytpes`.

**Comment by leofang** [comment]
@dalcinl We still need some `ctypes` machinery such as annotating pointers, so that we can e.g. register foreign C function signatures. What would be the alternative if we do not use `ctypes`?

**Comment by dalcinl** [comment]
Oh, ok, now I see what you mean, NumPy datatypes do not support typed pointers. I'm wondering whether you want to develop your own solution or DSL, perhaps based on simple strings with typecodes and the extension from the buffer interface for structs (`T{...}`) and a way to express pointers (e.g. a `*` suffix after a typecode). One thing I do not like about ctypes is that you cannot express ABIs and fixed-size types, that is, if you do `ctypes.c_int64`, that is an alias for `ctypes.c_long` or `ctypes.c_longlog` depending on the platform. Maybe this point is not really that important to CuPy.  But I still have the feeling that ctypes should not be at the core of the thing, at some point you may run short. Just my two cents...

**Comment by leofang** [comment]
I dunno why I dropped the ball here. Yeah we have a DSL in CuPy here so the idea would be to parse `ctypes.c_int64` and transpile it to `int64_t` so there's no ABI ambiguity. ctypes is only a way for us to express the type system, we likely won't use its implementation.

**Comment by dalcinl** [comment]
> the idea would be to parse `ctypes.c_int64` and transpile it to `int64_t`

That would not work in general, as `ctypes.c_int64` is just an alias for a platform-dependent type (`long` or `long long`).
In other words, there is no way to distinguish between `ctypes.c_int64` and `ctypes.c_long`/`ctypes.c_longlong`. 
Being able to distinguish these two types is important in the context of APIs/ABIs and platform independence.  
The `ctypes` type machinery does not allow for such distinction.
